Speed up prepare-features

oatsu-gh opened the pull request:
For a large singing-databese, preparing features can be the 1st or 2nd bottleneck in training. I made it faster (certified).

- Speed up prepare-features by multi-threading using `joblib.Parallel()`
- Use `logger.info(OmegaConf.to_yaml(config))` instead of `logger.info(config.pretty())` to avoid warnings
- Remove unused imports

r9y9 commented:
Sorry for the late reply. I have been busy at work. I will get to review once I get some time!

r9y9 commented on nnsvs/bin/prepare_features.py:
```diff
@@ -105,44 +156,40 @@ def my_app(config : DictConfig) -> None:
-            ypath = join(out_timelag_root, name + "-feats.npy")
-            np.save(xpath, x, allow_pickle=False)
-            np.save(ypath, y, allow_pickle=False)
+        joblib.Parallel(n_jobs=-2)(
+            joblib.delayed(_prepare_timelag_feature)(
+                in_timelag_root, out_timelag_root, in_timelag, out_timelag, idx
+            )
+            for idx in tqdm(range(len(in_duration)))
+        )
```
Could you replace `joblib.Parallel` to `concurrent.futures.ProcessPoolExecutor`? The later is more tqdm friendly. We cannot easily make tqdm work if we use joblib.Parallel.

You can refer to my another repository: https://github.com/r9y9/ttslearn/blob/23a4088571ac680f7e28b6ac59d3e2eb31b9511f/recipes/dnntts/preprocess_acoustic.py#L95-L110

r9y9 replied:
Also, it would be nice if we add `n_jobs` in config.yaml so that we can control it through the CLI (i.e., run.sh or prepare_features.py).

r9y9 commented on nnsvs/bin/prepare_features.py:
```diff
@@ -105,44 +156,40 @@ def my_app(config : DictConfig) -> None:
-            ypath = join(out_duration_root, name + "-feats.npy")
-            np.save(xpath, x, allow_pickle=False)
-            np.save(ypath, y, allow_pickle=False)
+        joblib.Parallel(n_jobs=-2)(
+            joblib.delayed(_prepare_duration_feature)(
+                in_duration_root, out_duration_root, in_duration, out_duration, idx
+            )
+            for idx in tqdm(range(len(in_duration)))
+        )
 
```
Same as above

r9y9 commented on nnsvs/bin/prepare_features.py:
```diff
@@ -105,44 +156,40 @@ def my_app(config : DictConfig) -> None:
-            np.save(xpath, x, allow_pickle=False)
-            np.save(ypath, y, allow_pickle=False)
-            np.save(wpath, wave, allow_pickle=False)
+        joblib.Parallel(n_jobs=-2)(
+            joblib.delayed(_prepare_acoustic_feature)(
+                in_acoustic_root, out_acoustic_root, in_acoustic, out_acoustic, idx
+            )
+            for idx in tqdm(range(len(in_duration)))
+        )
+
```
Same as above

r9y9 reviewed (changes requested):
Thanks for the PR!  I have two requests to make the code better. Please check my inline comments.

r9y9 commented:
BTW, we can ignore CI errors. I will fix pysinsy https://github.com/r9y9/pysinsy/issues/6 and CI accordingly.

r9y9 commented:
Done fixing the pysinsy and CI issues (https://github.com/r9y9/nnsvs/pull/51). You can fix CI by merging or rebasing on master.

oatsu-gh commented:
Sorry for having you wait so long. I've just learned about git rebasing. Now I'm learning `concurrent.futures.ProcessPoolExecutor` and developing the codes above. Thank you.

oatsu-gh commented:
誤って re-review をリクエストしてしまいました。無視していただけますと幸いです。

oatsu-gh commented:
- In prepare_features, Changed to use `ProcessPoolExecutor` with `config.max_workers`.
  - What do you think about the config name "max_workers"? Would it be better to use "n_jobs" or "num_workers" than "max_workers"?
- Added a new line `max_workers: none` into bin/conf/prepare_features/config.yaml.

r9y9 commented on nnsvs/bin/prepare_features.py:
```diff
@@ -96,53 +147,72 @@ def my_app(config : DictConfig) -> None:
-            np.save(xpath, x, allow_pickle=False)
-            np.save(ypath, y, allow_pickle=False)
-            np.save(wpath, wave, allow_pickle=False)
+    if config.timelag.enabled:
```
```suggestion
    if config.duration.enabled:
```

r9y9 commented on nnsvs/bin/prepare_features.py:
```diff
@@ -96,53 +147,72 @@ def my_app(config : DictConfig) -> None:
+                future.result()
+
+    # Save features for duration model
+    if config.timelag.enabled:
```
```suggestion
    if config.acoustic.enabled:
```

r9y9 commented on nnsvs/bin/prepare_features.py:
```diff
@@ -96,53 +147,72 @@ def my_app(config : DictConfig) -> None:
+            for future in tqdm(futures):
+                future.result()
+
+    # Save features for duration model
```
```suggestion
    # Save features for acoustic model
```

r9y9 reviewed (commented):
Added a few comments. The name `max_workers` looks good to me.

oatsu-gh commented:
Thank you for the checking and fixing!

r9y9 reviewed (approved):
LGTM, thank you!
